File: src/main/java/GuessNumberHigherOrLower/BBIGBOSS/GuessNumberHigherOrLower.cpp
//
// Created by BBIGBOSS on 2019-06-17.
//

// Forward declaration of guess API.
// @param num, your guess
// @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
int guess(int num);

class Solution
{
    public:

    int guessNumber(int n)
    {
        long left = 0, right = n;

        while (left < right)
        {
            long mid = (left + right) / 2;

            if (guess(mid) == 1)
                left = mid + 1;
            else if (guess(mid) == -1)
                right = mid;
            else
                return mid;
        }

        return left;
    }
};

File: src/main/java/GuessNumberHigherOrLower/haierbopu/GuessNumberHigherOrLower.java
/* The guess API is defined in the parent class GuessGame.
   @param num, your guess
   @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
      int guess(int num); */

public class Solution extends GuessGame {
    public int guessNumber(int n) {
        int l = 1;
        int r = n;
        //still same binary search like problem 1, but one more condition for correct answer
        while (l < r){
            int temp = (r - l)/2 + l;
            if (guess(temp) == 0){
                return temp;
            }
            else if (guess(temp) < 0){
                r = temp;
            }
            else{
                l = temp + 1;
            }
        }
        return l;
    }
}